include/xgboost/predictor.h

```diff
@@ -186,41 +170,14 @@ class Predictor {
   static Predictor* Create(std::string name);
 
  protected:
-  /**
-   * \fn  bool PredictFromCache(DMatrix* dmat, std::vector<bst_float>*
-   * out_preds, const gbm::GBTreeModel& model, unsigned ntree_limit = 0)
-   *
-   * \brief Attempt to predict from cache.
-   *
-   * \return  True if it succeeds, false if it fails.
-   */
-  bool PredictFromCache(DMatrix* dmat, std::vector<bst_float>* out_preds,
-                        const gbm::GBTreeModel& model,
-                        unsigned ntree_limit = 0);
-
```

Comment:
Currently, these functions are defined in the Predictor class, but used only in CPUPredictor. GPUPredictor defines and uses its own versions of these functions. Since these functions are not virtual, the implementation in GPUPredictor does not override them, but just hides them, and I wanted to avoid that. Therefore, I moved these functions from Predictor to CPUPredictor, where they are actually used.

The alternatives considered:

- define these functions in GPUPredictor under a different name; I also wanted to avoid that, since the GPUPredictor does not use functions defined in Predictor;
- make these functions virtual, together with DevicePredictInternal/PredLoopInternal, and have the common logic in Predictor; this, however, quickly amounts to a refactoring of the predictors, and is a topic for a different pull request.

I'm happy to move these functions back to Predictor and have the GPU-based versions of these functions named differently in GPUPredictor, if you actually prefer it that way.

Reply:
Let's just remove these functions from Predictor class entirely, as they seem to be more of an implementation detail rather than a public interface. We can do another PR in the future if the functions need to be exposed again.